File: apac_kiosk/monitoring/screen_capture.py
from __future__ import annotations

import os
import threading
import time
from datetime import datetime
from PIL import ImageGrab, Image
import io

from ..database import db


class ScreenCapture:
    def __init__(self):
        self._running = False
        self._thread: threading.Thread | None = None
        self._user_id: int | None = None
        self._username: str = ""

    def start(self, user_id: int, username: str):
        self._user_id = user_id
        self._username = username
        self._running = True
        self._thread = threading.Thread(target=self._capture_loop, daemon=True)
        self._thread.start()

    def stop(self):
        self._running = False

    def _capture_loop(self):
        while self._running:
            try:
                interval = int(db.get_config("screenshot_interval", "60"))
                self._capture_once()
            except Exception:
                pass
            time.sleep(interval)

    def _capture_once(self):
        try:
            img = ImageGrab.grab(all_screens=True)
            quality = db.get_config("screenshot_quality", "medium")
            quality_map = {"high": 95, "medium": 70, "low": 40}
            jpg_quality = quality_map.get(quality, 70)

            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            filename = f"screenshot_{timestamp}_{self._username}.jpg"
            folder = db.get_config("screenshot_path", os.path.join(os.path.dirname(db.DB_PATH), "screenshots"))
            os.makedirs(folder, exist_ok=True)
            filepath = os.path.join(folder, filename)

            if quality == "low":
                img = img.resize((img.width // 2, img.height // 2), Image.LANCZOS)

            img.save(filepath, "JPEG", quality=jpg_quality)
            db.insert_log(
                self._user_id, "screenshot", file_path=filepath,
                description=f"Captura de tela"
            )
        except Exception:
            pass

File: ApacKiosk/Services/ScreenCapture.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Threading;
using ApacKiosk.Database;
using ApacKiosk.Utils;

namespace ApacKiosk.Services
{
    public class ScreenCapture
    {
        private readonly ConfigManager _config;
        private readonly DatabaseManager _db;
        private Thread _thread;
        private volatile bool _isRunning;
        private int? _currentUserId;

        public ScreenCapture(ConfigManager config, DatabaseManager db)
        {
            _config = config;
            _db = db;
        }

        public void SetCurrentUser(int? userId)
        {
            _currentUserId = userId;
        }

        public void Start()
        {
            if (_isRunning) return;
            _isRunning = true;
            _thread = new Thread(CaptureLoop)
            {
                IsBackground = true,
                Name = "ScreenCapture"
            };
            _thread.Start();
        }

        public void Stop()
        {
            _isRunning = false;
            _thread?.Join(3000);
        }

        private void CaptureLoop()
        {
            while (_isRunning)
            {
                try
                {
                    if (_config.ScreenshotEnabled && _currentUserId.HasValue)
                    {
                        var interval = Math.Max(_config.ScreenshotIntervalSec, 10);
                        CaptureNow();
                    }
                }
                catch (Exception ex)
                {
                    _db.InsertLog(_currentUserId, "screenshot_error", null, ex.Message);
                }

                var sleepTime = Math.Max(_config.ScreenshotIntervalSec, 10) * 1000;
                var slept = 0;
                while (slept < sleepTime && _isRunning)
                {
                    Thread.Sleep(100);
                    slept += 100;
                }
            }
        }

        public void CaptureNow()
        {
            var userLabel = _currentUserId?.ToString() ?? "system";
            var timestamp = DateTime.Now.ToString("yyyyMMdd_HHmmss");
            var dir = _config.ScreenshotPath;
            Directory.CreateDirectory(dir);

            var fileName = $"screenshot_{timestamp}_{userLabel}.jpg";
            var filePath = Path.Combine(dir, fileName);

            using var bmp = new Bitmap(System.Windows.Forms.Screen.PrimaryScreen.Bounds.Width,
                                       System.Windows.Forms.Screen.PrimaryScreen.Bounds.Height);
            using var g = Graphics.FromImage(bmp);
            g.CopyFromScreen(0, 0, 0, 0, bmp.Size);

            int quality;
            switch (_config.ScreenshotQuality)
            {
                case "High": quality = 90; break;
                case "Low": quality = 40; break;
                default: quality = 70; break;
            }

            var encoder = ImageCodecInfo.GetImageEncoders();
            var jpegEncoder = Array.Find(encoder, e => e.MimeType == "image/jpeg");
            var encoderParams = new EncoderParameters(1);
            encoderParams.Param[0] = new EncoderParameter(System.Drawing.Imaging.Encoder.Quality, quality);

            bmp.Save(filePath, jpegEncoder, encoderParams);
            encoderParams.Dispose();

            _db.InsertLog(_currentUserId, "screenshot", filePath, null);
        }
    }
}
